test/modules/TestResultCoordinator/Reports/EdgeHubRestartTest/EdgeHubRestartDirectMethodReportGenerator.cs

```diff
@@ -0,0 +1,304 @@
+                resultCount,
+                completedStatusHistogram);
+
+            return this.CalculateStatistic(
+                isPassing,
+                passedDirectMethodCount,
+                resultCount,
+                completedStatusHistogram);
+        }
+
+        EdgeHubRestartDirectMethodReport CalculateStatistic(
+            bool isPassing,
+            long passedDirectMethodCount,
+            Dictionary<string, ulong> resultCount,
+            Dictionary<HttpStatusCode, List<TimeSpan>> completedStatusHistogram)
+        {
+            List<TimeSpan> completedPeriods;
+            completedStatusHistogram.TryGetValue(HttpStatusCode.OK, out completedPeriods);
+            List<TimeSpan> orderedCompletedPeriods = completedPeriods?.OrderBy(p => p.Ticks).ToList();
+
+            TimeSpan minPeriod = TimeSpan.FromTicks(0);
+            TimeSpan maxPeriod = TimeSpan.FromTicks(0);
+            TimeSpan medianPeriod = TimeSpan.FromTicks(0);
+            TimeSpan meanPeriod = TimeSpan.FromTicks(0);
+            TimeSpan variancePeriod = TimeSpan.FromTicks(0);
+            if (orderedCompletedPeriods != null)
+            {
+                minPeriod = orderedCompletedPeriods.First();
+                maxPeriod = orderedCompletedPeriods.Last();
+
+                if ((orderedCompletedPeriods.Count & 0b1) == 0b1)
+                {
+                    // If odd, pick the middle value
+                    medianPeriod = orderedCompletedPeriods[orderedCompletedPeriods.Count >> 1];
+                }
+                else
+                {
+                    // If even, average the middle values
+                    medianPeriod =
+                        (orderedCompletedPeriods[orderedCompletedPeriods.Count >> 1] +
+                        orderedCompletedPeriods[(orderedCompletedPeriods.Count >> 1) - 1]) / 2;
+                }
+
+                // Compute Mean
+                TimeSpan totalSpan = TimeSpan.FromTicks(0);
+                double totalSpanSquareInMilisec = 0.0;
+                foreach (TimeSpan eachTimeSpan in orderedCompletedPeriods)
+                {
+                    totalSpan += eachTimeSpan;
+                    totalSpanSquareInMilisec += Math.Pow(eachTimeSpan.TotalMilliseconds, 2);
+                }
+
+                // Compute Mean : mean = sum(x) / N
+                meanPeriod = totalSpan / orderedCompletedPeriods.Count();
+
+                // Compute Variance: var = sum((x - mean)^2) / N
+                //                       = sum(x^2) / N - mean^2
+                double variancePeriodInMilisec = (totalSpanSquareInMilisec / orderedCompletedPeriods.Count()) - Math.Pow(meanPeriod.TotalMilliseconds, 2);
+                variancePeriod = TimeSpan.FromMilliseconds(variancePeriodInMilisec);
+            }
+
+            // Make sure the maximum restart period is within a passable threshold
+            isPassing &= maxPeriod < this.Metadata.PassableEdgeHubRestartPeriod;
```

Comment:
why do we need this?  this is doing similar thing as what runExpirationTime during each edgeHub restart test run.

Reply:
I see what you mean. 
I'll keep this param for Direct Method, but remove it for Message. 
For Message, I'll just check the complete HTTP status if it's passing.